handler/rfc8693/custom_jwt_type_handler.go

```diff
@@ -0,0 +1,238 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package rfc8693
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/ory/fosite"
+	"github.com/ory/fosite/handler/openid"
+	"github.com/ory/fosite/token/jwt"
+	"github.com/ory/x/errorsx"
+)
+
+type CustomJWTTypeHandler struct {
+	Config      fosite.RFC8693ConfigProvider
+	JWTStrategy jwt.Signer
+	Storage
+}
+
+// HandleTokenEndpointRequest implements https://tools.ietf.org/html/rfc6749#section-4.3.2
+func (c *CustomJWTTypeHandler) HandleTokenEndpointRequest(ctx context.Context, request fosite.AccessRequester) error {
+	if !c.CanHandleTokenEndpointRequest(ctx, request) {
+		return errorsx.WithStack(fosite.ErrUnknownRequest)
+	}
+
+	session, _ := request.GetSession().(Session)
+	if session == nil {
+		return errorsx.WithStack(fosite.ErrServerError.WithDebug("Failed to perform token exchange because the session is not of the right type."))
+	}
+
+	form := request.GetRequestForm()
+	tokenTypes := c.Config.GetTokenTypes(ctx)
+	actorTokenType := tokenTypes[form.Get("actor_token_type")]
+	subjectTokenType := tokenTypes[form.Get("subject_token_type")]
+	if actorTokenType != nil && actorTokenType.GetType(ctx) == JWTTokenType {
+		token := form.Get("actor_token")
+		if unpacked, err := c.validate(ctx, request, actorTokenType, token); err != nil {
+			return err
+		} else {
+			session.SetActorToken(unpacked)
+		}
+	}
+
+	if subjectTokenType != nil && subjectTokenType.GetType(ctx) == JWTTokenType {
+		token := form.Get("subject_token")
+		if unpacked, err := c.validate(ctx, request, subjectTokenType, token); err != nil {
+			return err
+		} else {
+			session.SetSubjectToken(unpacked)
+			// Get the subject and populate session
+			if subject, err := c.Storage.GetSubjectForTokenExchange(ctx, request, unpacked); err != nil {
+				return err
+			} else {
+				session.SetSubject(subject)
+			}
+		}
+	}
+
+	return nil
+}
+
+// PopulateTokenEndpointResponse implements https://tools.ietf.org/html/rfc6749#section-4.3.3
+func (c *CustomJWTTypeHandler) PopulateTokenEndpointResponse(ctx context.Context, request fosite.AccessRequester, responder fosite.AccessResponder) error {
+
+	if !c.CanHandleTokenEndpointRequest(ctx, request) {
+		return errorsx.WithStack(fosite.ErrUnknownRequest)
+	}
+
+	session, _ := request.GetSession().(Session)
+	if session == nil {
+		return errorsx.WithStack(fosite.ErrServerError.WithDebug("Failed to perform token exchange because the session is not of the right type."))
+	}
+
+	form := request.GetRequestForm()
+	requestedTokenType := form.Get("requested_token_type")
+	if requestedTokenType == "" {
+		requestedTokenType = c.Config.GetDefaultRequestedTokenType(ctx)
+	}
+
+	tokenTypes := c.Config.GetTokenTypes(ctx)
+	tokenType := tokenTypes[requestedTokenType]
+	if tokenType == nil || tokenType.GetType(ctx) != JWTTokenType {
+		return nil
+	}
+
+	if err := c.issue(ctx, request, tokenType, responder); err != nil {
+		return err
+	}
+
```

Comment:
should we add `issued_token_type` https://github.ibm.com/strata/ristretto-core/blob/master/fositex/handler/rfc8693/custom_jwt_type_handler.go#L95